File: app/src/main/java/com/zeyad/rxredux/screens/BaseActivity.java
package com.zeyad.rxredux.screens;

import java.util.List;

import com.zeyad.rxredux.core.redux.BaseViewModel;
import com.zeyad.rxredux.snackbar.SnackBarFactory;

import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentTransaction;
import android.util.Pair;
import android.view.View;
import android.widget.Toast;

/**
 * @author by ZIaDo on 7/21/17.
 */
public abstract class BaseActivity<S, VM extends BaseViewModel<S>>
        extends com.zeyad.rxredux.core.redux.prelollipop.BaseActivity<S, VM> {

    /**
     * Adds a {@link Fragment} to this activity's layout.
     *
     * @param containerViewId The container view to where add the fragment.
     * @param fragment        The fragment to be added.
     */
    public void addFragment(int containerViewId, Fragment fragment, String currentFragTag,
            List<Pair<View, String>> sharedElements) {
        FragmentTransaction fragmentTransaction = getSupportFragmentManager().beginTransaction();
        if (sharedElements != null) {
            for (Pair<View, String> pair : sharedElements) {
                fragmentTransaction.addSharedElement(pair.first, pair.second);
            }
        }
        if (currentFragTag == null || currentFragTag.isEmpty()) {
            fragmentTransaction.addToBackStack(fragment.getTag());
        } else {
            fragmentTransaction.addToBackStack(currentFragTag);
        }
        fragmentTransaction.add(containerViewId, fragment, fragment.getTag()).commit();
    }

    public void removeFragment(String tag) {
        getSupportFragmentManager().beginTransaction().remove(getSupportFragmentManager().findFragmentByTag(tag))
                .commit();
    }

    public void showToastMessage(String message) {
        showToastMessage(message, Toast.LENGTH_LONG);
    }

    public void showToastMessage(String message, int duration) {
        Toast.makeText(this, message, duration).show();
    }

    /**
     * Shows a {@link android.support.design.widget.Snackbar} message.
     *
     * @param message An string representing a message to be shown.
     */
    public void showSnackBarMessage(View view, String message, int duration) {
        if (view != null) {
            SnackBarFactory.getSnackBar(SnackBarFactory.TYPE_INFO, view, message, duration).show();
        } else {
            throw new IllegalArgumentException("View is null");
        }
    }

    public void showSnackBarWithAction(String typeSnackBar, View view, String message, String actionText,
            View.OnClickListener onClickListener) {
        if (view != null) {
            SnackBarFactory.getSnackBarWithAction(typeSnackBar, view, message, actionText, onClickListener).show();
        } else {
            throw new IllegalArgumentException("View is null");
        }
    }

    public void showSnackBarWithAction(String typeSnackBar, View view, String message, int actionText,
            View.OnClickListener onClickListener) {
        showSnackBarWithAction(typeSnackBar, view, message, getString(actionText), onClickListener);
    }

    /**
     * Shows a {@link android.support.design.widget.Snackbar} errorResult message.
     *
     * @param message  An string representing a message to be shown.
     * @param duration Visibility duration.
     */
    public void showErrorSnackBar(String message, View view, int duration) {
        if (view != null) {
            SnackBarFactory.getSnackBar(SnackBarFactory.TYPE_ERROR, view, message, duration).show();
        } else {
            throw new IllegalArgumentException("View is null");
        }
    }
}

File: app/src/main/java/com/zeyad/rxredux/screens/BaseActivity.kt
package com.zeyad.rxredux.screens

import android.os.Parcelable
import android.support.v4.app.Fragment
import android.util.Pair
import android.view.View
import android.widget.Toast
import com.zeyad.rxredux.core.view.BaseActivity
import com.zeyad.rxredux.core.viewmodel.BaseViewModel
import com.zeyad.rxredux.snackbar.SnackBarFactory

/**
 * @author by ZIaDo on 7/21/17.
 */
abstract class BaseActivity<S : Parcelable, VM : BaseViewModel<S>> : BaseActivity<S, VM>() {

    /**
     * Adds a [Fragment] to this activity's layout.
     *
     * @param containerViewId The container view to where add the fragment.
     * @param fragment        The fragment to be added.
     */
    @SafeVarargs
    fun addFragment(containerViewId: Int, fragment: Fragment, currentFragTag: String?,
                    vararg sharedElements: Pair<View, String>) {
        val fragmentTransaction = supportFragmentManager.beginTransaction()
        for (pair in sharedElements) {
            fragmentTransaction.addSharedElement(pair.first, pair.second)
        }
        if (currentFragTag == null || currentFragTag.isEmpty()) {
            fragmentTransaction.addToBackStack(fragment.tag)
        } else {
            fragmentTransaction.addToBackStack(currentFragTag)
        }
        fragmentTransaction.add(containerViewId, fragment, fragment.tag).commit()
    }

    fun removeFragment(tag: String) {
        supportFragmentManager.beginTransaction().remove(supportFragmentManager.findFragmentByTag(tag))
                .commit()
    }

    @JvmOverloads
    fun showToastMessage(message: String, duration: Int = Toast.LENGTH_LONG) {
        Toast.makeText(this, message, duration).show()
    }

    /**
     * Shows a [android.support.design.widget.Snackbar] message.
     *
     * @param message An string representing a message to be shown.
     */
    fun showSnackBarMessage(view: View?, message: String, duration: Int) {
        if (view != null) {
            SnackBarFactory.getSnackBar(SnackBarFactory.TYPE_INFO, view, message, duration).show()
        } else {
            throw IllegalArgumentException("View is null")
        }
    }

    fun showSnackBarWithAction(typeSnackBar: String, view: View?, message: String, actionText: String,
                               onClickListener: View.OnClickListener) {
        if (view != null) {
            SnackBarFactory.getSnackBarWithAction(typeSnackBar, view, message, actionText, onClickListener).show()
        } else {
            throw IllegalArgumentException("View is null")
        }
    }

    fun showSnackBarWithAction(typeSnackBar: String, view: View, message: String, actionText: Int,
                               onClickListener: View.OnClickListener) {
        showSnackBarWithAction(typeSnackBar, view, message, getString(actionText), onClickListener)
    }

    /**
     * Shows a [android.support.design.widget.Snackbar] errorResult message.
     *
     * @param message  An string representing a message to be shown.
     * @param duration Visibility duration.
     */
    fun showErrorSnackBar(message: String, view: View?, duration: Int) {
        if (view != null) {
            SnackBarFactory.getSnackBar(SnackBarFactory.TYPE_ERROR, view, message, duration).show()
        } else {
            throw IllegalArgumentException("View is null")
        }
    }
}
